src/framework/enhancement/js/ContextAwareness.js

```diff
@@ -0,0 +1,253 @@
+        var root = segs.length === 0;
+        if (root) {
+            gradeNames.push("fluid.resolveRoot");
+        }
+        var memberName = fluid.typeNameToMemberName(options.singleRootType || type);
+        segs.push(memberName);
+        fluid.construct(segs, options, instantiator);
+    };
+    
+    /** Destroy an instance created by `fluid.constructSingle`
+     * @param parentPath {String|Array of String} Parent of path where the new component is to be constructed, represented as a string or array of segments
+     * @param typeName {String} The type name used to construct the component (either `type` or `singleRootType` of the `options` argument to `fluid.constructSingle`
+     * @param instantiator {Instantiator} [optional] The instantiator holding the component to be created - if blank, the global instantiator will be used
+    */
+    fluid.destroySingle = function (parentPath, typeName, instantiator) {
+        instantiator = instantiator || fluid.globalInstantiator;
+        var segs = fluid.model.parseToSegments(parentPath, instantiator.parseEL, true);
+        var memberName = fluid.typeNameToMemberName(typeName);
+        segs.push(memberName);
+        fluid.destroy(segs, instantiator);
+    };
+    
+    fluid.defaults("fluid.contextAware.marker", {
+        gradeNames: ["fluid.component"]
+    });
+    
+    
+    // unsupported, NON-API function
+    fluid.contextAware.makeCheckMarkers = function (checks, path, instantiator) {
+        fluid.each(checks, function (value, markerTypeName) {
+            fluid.constructSingle(path, {
+                type: markerTypeName,
+                gradeNames: "fluid.contextAware.marker",
+                value: value
+            }, instantiator);
+        });
+        
+    };
+    /** Peforms the computation for `fluid.contextAware.makeChecks` and returns a hash suitable for being sent to `fluid.contextAware.makeCheckMarkers` - 
+     *
+     * @return A hash of horizon names to grade names - this can be sent to fluid.contextAware.makeCheckMarkers
+     */
+    // unsupported, NON-API function
+    fluid.contextAware.performChecks = function (checkHash) {
+        return fluid.transform(checkHash, function (checkRecord) {
+            if (fluid.isPrimitive(checkRecord)) {
+                return checkRecord;
+            } else if ("value" in checkRecord) {
+                return checkRecord.value;
+            } else if ("func" in checkRecord) {
+                return checkRecord.func();
+            } else if ("funcName" in checkRecord) {
+                return fluid.invokeGlobalFunction(checkRecord.funcName);
+            } else {
+                fluid.fail("Error in contextAwareness check record ", checkRecord, " - must contain an entry with name value, func, or funcName");
+            }
+        });
+    };
+
+    /**
+     * Takes an hash of check context names to check records, designating a collection of context markers which might be registered at a location
```

Comment:
I'd prefer that we avoid the term "hash" here (and else where in this file) in favour of "object" or "structure" or "specification"--perhaps even consistently naming the this kind of specification explicitly--"contextCheckSpec" or something like that. As a new reader (especially native JavaScripters), the "hash" terminology is confusing.

I can imagine that our documentation, in the future, may be quite careful about defining these types of declarative specifications up front (e.g. [we do a nice job for invoker specs](http://docs.fluidproject.org/infusion/development/Invokers.html#standard-invoker-binding-to-a-function-using-funcname-func) or whatever they should be called).